### PR TITLE
Improve log flexibility in HEMCO

### PR DESCRIPTION
**Pull request by lizziel**
### Name and Institution (Required)

Name: Lizzie Lundgren
Institution: Harvard University

### Describe the update

This is a zero diff PR that does a structural overhaul of error and log handling in HEMCO. Its motivation is to enable writing HEMCO log messages to an existing external model's log already open at the time of HEMCO initialization and which is not used for standard output. The use case for this is sending all HEMCO log messages to the CESM CAM log `atm.log`. Without these updates HEMCO prints go to `cesm.log` which is a high-level log file not appropriate for HEMCO messages unless error.

This PR also contains a lot of cleanup of the log and error handling which hopefully improves control over logging. The primary changes are removing the Err object from `HCO_ERROR`, `HCO_MSG`, and `HCO_WARNING`, and updating those to be subroutines (not interfaces) that are (1) independent of verbose and core (e.g. if root), and (2) write to stdout unless passed optional argument LUN. This allows control of when and where messages are printed on a case-by-case basis.

Specific changes in this PR (based on the changelog):

Added:
- Added optional argument `LUN` (log file logical unit number) to `ConfigInit` to allow external models to pass LUN of existing log file
- Added two log LUN variables (`stdLogLUN` and `hcoLogLUN`) to `HcoState%Config` for stdout and HEMCO log and initialize both in `ConfigInit` to stdout or optional LUN (if passed)
- Added output argument `LUN` to `HCO_LOGFILE_OPEN` to update `HcoState%Config%hcoLogLun` to LUN of new log (if using)
- Added optional LUN argument to `HCO_MSG`, `HCO_WARNING`, and `HCO_ERROR` to specify log to print to
- Added special log handling for CESM to ignore `LogFile` entry in `HEMCO_Config.rc` (will get passed CAM log LUN to use instead)

Changed:
- Changed interfaces `HCO_MSG`, `HCO_WARNING`, and `HCO_ERROR` to each be a single subroutine that does not use `HcoState%Config%Err`
- Changed `HCO_MSG`, `HCO_WARNING`, and `HCO_ERROR` to be independent of verbose and cores
- Updated calls to `HCO_MSG` to send output to `HcoState%Config%hcoLogLUN`
- Updated calls to `HCO_WARNING` to print only if verbose and to send them to stdout unless related to units. Units warnings can mostly be ignored since we use unit `1` in `HEMCO_Config.rc` to instruct HEMCO not to convert units but do not disable a warning about a unit mismatch between the configuration file and data file.
- Replaced usage of `HCO_IsVerb` with `HcoState_Config%doVerbose` to reduce reliance on the `Err` object for prints
- Updated documentation in `HCO_Error_Mod.F90` to summarize error and log handling in HEMCO

Fixed:
- Fixed excessive prints when using MPI

Removed:
- Removed warnings count in `HcoState%Config%Err` since not useful (see comment above about unit mismatch warnings)
- Removed RC argument in `HCO_WARNING`
- Deleted subroutine `HCO_IsVerb`

### Expected changes

This is a zero diff update for all simulations. 

For this update I tested with GC-Classic, GCHP, and HEMCO standalone. I inspected logs files with and without verbose turned on, and with and without using a HEMCO.log file. There are several minor tweaks to print calls based on what I saw to make non-verbose readability good. A good sanity check tool is now to turn on HEMCO verbose and enable HEMCO.log. This will print warnings of potential concern to the main GEOS-Chem log and everything else to HEMCO.log.

### Reference(s)

None

### Related Github Issue

None

### CODE REVIEW

File: src/Core/hco_error_mod.F90 [hunk 1]
```diff
@@ -6,47 +6,74 @@
...
-! As of HEMCO v2.0, all HEMCO error variables are organized in derived
-! type object HcoErr. HcoErr is a component of the HEMCO configuration
-! object (type ConfigObj, see hco\_types\_mod.F90). It must be passed
-! explicitly to all error routines. This design allows the invocation
-! of multiple independent HEMCO instances at the same time (which may
-! have different HEMCO error settings).
+! for error handling and logfile messages in HEMCO. It alse contains
+! definitions of some globally used parameters, such as single/double
+! precision as well as the HEMCO precision which can be either single or
+! couble. The HEMCO precision is used for almost all HEMCO internal data
```
Comment (yantosca) [inline src/Core/hco_error_mod.F90:12]: Typo `couble` -> `double`

File: src/Core/hco_error_mod.F90 [hunk 2]
```diff
@@ -6,47 +6,74 @@
...
-!\\
-!\\
-! As of HEMCO v2.0, all HEMCO error variables are organized in derived
-! type object HcoErr. HcoErr is a component of the HEMCO configuration
-! object (type ConfigObj, see hco\_types\_mod.F90). It must be passed
-! explicitly to all error routines. This design allows the invocation
-! of multiple independent HEMCO instances at the same time (which may
-! have different HEMCO error settings).
+! for error handling and logfile messages in HEMCO. It alse contains
+! definitions of some globally used parameters, such as single/double
```
Comment (yantosca) [inline src/Core/hco_error_mod.F90:10]: @lizziel: We should also update the ReadTheDocs accordingly in this section:
https://hemco.readthedocs.io/en/stable/hco-ref-guide/hemco-config.html#logfile

This can also be done as a separate commit to the `docs/dev` branch.

File: src/Core/hcoio_read_std_mod.F90 [hunk 3]
```diff
@@ -1131,35 +1134,41 @@ SUBROUTINE HCOIO_Read( HcoState, Lct, RC )
...
        ENDIF
 
     ! Convert to HEMCO units in all other cases.
     ELSE
 
        ! For zero unit tolerance, make sure that thisUnit matches
        ! with unit set in configuration file. For higher unit
-       ! tolerances, prompt a level 3 warning.
+       ! tolerances, prompt a level 3 warning. Only print the warning
+       ! if verbose since most instances of unit mismatches are normal,
```
Comment (yantosca) [inline src/Core/hcoio_read_std_mod.F90:1160]: @lizziel: Change `level 3 warning` to `warning`.  The level 3 was a holdover from the numeric verbose values.

**Review by yantosca — CHANGES_REQUESTED** [review]
@lizziel you win the award for biggest PR ever!  :-)

I just noticed a few minor things that needed fixing.  Also we'll need to update the RTD documentation as noted.   See my comments.

File: src/Core/hco_error_mod.F90 (hunk 1 above)
Comment (lizziel) [reply src/Core/hco_error_mod.F90]: Now fixed.

File: src/Core/hco_error_mod.F90 (hunk 2 above)
Comment (lizziel) [reply src/Core/hco_error_mod.F90:10]: Will do!

File: src/Core/hcoio_read_std_mod.F90 (hunk 3 above)
Comment (lizziel) [reply src/Core/hcoio_read_std_mod.F90:1160]: Now fixed.

**Review by yantosca — APPROVED** [review]
Thanks @lizziel!  I approve conditionally pending the doc update.

File: src/Core/hcoio_read_std_mod.F90 [hunk 4]
```diff
@@ -194,6 +194,7 @@ SUBROUTINE HCOIO_Read( HcoState, Lct, RC )
     LOGICAL                       :: FOUND
     LOGICAL                       :: IsModelLevel
     LOGICAL                       :: DoReturn
+    LOGICAL                       :: IGNORE
```
Comment (msulprizio) [inline src/Core/hcoio_read_std_mod.F90:197]: Not sure why this was added. It doesn't seem to be used.

**Review by msulprizio — APPROVED** [review]
Thanks @lizziel. I have only one minor question/comment but otherwise this looks good to merge.

File: src/Core/hcoio_read_std_mod.F90 (hunk 4 above)
Comment (lizziel) [reply src/Core/hcoio_read_std_mod.F90:197]: Good catch. I think I was going to add something to ignore certain warnings, specifically the onces about data being treated as unitless but units are present in the file, e.g. mol mol-1 for restart variables. But then I decided against it, at least in this update.

Comment (msulprizio) [reply src/Core/hcoio_read_std_mod.F90]: Resolved by https://github.com/geoschem/HEMCO/pull/298/commits/428c58da2dfced0a48f564504477a0e14249a93b
